lib/ACL/ACLManager.php

```diff
@@ -101,18 +103,33 @@ private function getParents(string $path): array {
 		return $paths;
 	}
 
-	public function preloadPaths(array $paths): void {
+	/**
+	 * @return Rule[][]
+	 */
+	public function preloadPaths(array $paths, bool $cache = true): array {
 		$allPaths = [];
 		foreach ($paths as $path) {
 			$allPaths = array_unique(array_merge($allPaths, $this->getParents($path)));
 		}
-		$this->getRules($allPaths);
+		return $this->getRules($allPaths, $cache);
 	}
 
 	public function getACLPermissionsForPath(string $path): int {
 		$path = ltrim($path, '/');
 		$rules = $this->getRules($this->getParents($path));
 
+		return $this->calculatePermissionsForPath($path, $rules);
+	}
+
+	public function getPermissionsForPathFromRules(string $path, array $rules): int {
+		$path = ltrim($path, '/');
+		$parents = $this->getParents($path);
+		// filter to only the rules we care about
+		$rules = $nonCachedPaths = array_intersect_key($rules, array_flip($parents));
+		return $this->calculatePermissionsForPath($path, $rules);
```

Comment:
`$nonCachedPaths` seems unused?

Reply:
Also this intersect_key suggests that `$rules` is indexed by path, so has string keys, which is not documented as phpdoc. It would help understand the logic if `$rules` and `$parents` type were better defined.

Reply:
Renamed "parent" to "relevant paths" and added some comments in various places to hopefully make the whole bit easier to understand.